code_folder/22_data_extraction_pipeline.py

```
@@ -0,0 +1,232 @@
+
+
+# ---------------------------------------------------------------------------
+# 5. Store and Retrieve from SQLite
+# ---------------------------------------------------------------------------
+def round_trip_sqlite(df: pd.DataFrame, table_name: str, db_path: str) -> pd.DataFrame:
+    """Write a dataframe to SQLite and read it back — simulates DB extraction."""
+    print(f"[Extract] SQLite round-trip: writing {len(df)} rows to '{table_name}'")
+    conn = sqlite3.connect(db_path)
+    try:
+        df.to_sql(table_name, conn, if_exists="replace", index=False)
+        df_back = pd.read_sql(f"SELECT * FROM {table_name}", conn)
```

Ruff (0.14.14)

[error] Possible SQL injection vector through string-based query construction

(S608)